File: python/src/pools/fixed_price_lbp/fixed_price_lbp_data.py
from dataclasses import dataclass, fields
from typing import List

from src.common.base_pool_state import BasePoolState


@dataclass
class FixedPriceLBPMutable:
    is_swap_enabled: bool
    current_timestamp: int


@dataclass
class FixedPriceLBPImmutable:
    project_token_index: int
    reserve_token_index: int
    project_token_rate: int
    start_time: int
    end_time: int


@dataclass
class FixedPriceLBPState(BasePoolState, FixedPriceLBPMutable, FixedPriceLBPImmutable):
    def __init__(self, **kwargs):
        kwargs["pool_type"] = "FIXED_PRICE_LBP"
        base_fields = {f.name for f in fields(BasePoolState)}
        base_kwargs = {k: v for k, v in kwargs.items() if k in base_fields}
        super().__init__(**base_kwargs)

        mutable_fields = {f.name for f in fields(FixedPriceLBPMutable)}
        immutable_fields = {f.name for f in fields(FixedPriceLBPImmutable)}

        for field in mutable_fields:
            if field in kwargs:
                setattr(self, field, kwargs[field])
        for field in immutable_fields:
            if field in kwargs:
                setattr(self, field, kwargs[field])


def map_fixed_price_lbp_state(pool_state: dict) -> FixedPriceLBPState:
    return FixedPriceLBPState(
        pool_address=pool_state["poolAddress"],
        tokens=pool_state["tokens"],
        scaling_factors=pool_state["scalingFactors"],
        token_rates=pool_state["tokenRates"],
        balances_live_scaled18=pool_state["balancesLiveScaled18"],
        swap_fee=pool_state["swapFee"],
        aggregate_swap_fee=pool_state.get("aggregateSwapFee", 0),
        total_supply=pool_state["totalSupply"],
        supports_unbalanced_liquidity=pool_state.get(
            "supportsUnbalancedLiquidity", False
        ),
        hook_type=pool_state.get("hookType", None),
        project_token_index=pool_state["projectTokenIndex"],
        reserve_token_index=pool_state["reserveTokenIndex"],
        project_token_rate=pool_state["projectTokenRate"],
        start_time=pool_state["startTime"],
        end_time=pool_state["endTime"],
        is_swap_enabled=pool_state["isSwapEnabled"],
        current_timestamp=pool_state["currentTimestamp"],
    )

File: rust/src/pools/fixed_price_lbp/fixed_price_lbp_data.rs
use crate::common::types::BasePoolState;
use alloy_primitives::U256;
use serde::{Deserialize, Serialize};

/// FixedPriceLBP mutable state
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FixedPriceLBPMutable {
    pub is_swap_enabled: bool,
    pub current_timestamp: U256,
}

/// FixedPriceLBP immutable state
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FixedPriceLBPImmutable {
    pub project_token_index: usize,
    pub reserve_token_index: usize,
    pub project_token_rate: U256,
    pub start_time: U256,
    pub end_time: U256,
}

/// FixedPriceLBP pool state
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FixedPriceLBPState {
    pub base: BasePoolState,
    pub mutable: FixedPriceLBPMutable,
    pub immutable: FixedPriceLBPImmutable,
}

impl From<FixedPriceLBPState> for crate::common::types::PoolState {
    fn from(state: FixedPriceLBPState) -> Self {
        crate::common::types::PoolState::FixedPriceLBP(state)
    }
}
